Add blog post and cover photo for 4.3 beta 1 (#856)
godotengine/godot-website: collections/_article/dev-snapshot-godot-4-3-beta-1.md

* Unbind GDExtension methods that can't reasonably be used ([GH-88418](https://github.com/godotengine/godot/pull/88418)).
* Fix some AcceptDialog argument types. Previous erroneous calls will now error instead of do nothing ([GH-89419](https://github.com/godotengine/godot/pull/89419)).

### 2D

#### Huge improvement to pixel stability for pixel art games
After much community discussion and the hard work of several contributors, we have merged a [PR](https://github.com/godotengine/godot/pull/87297) that we think resolves many of the outstanding issues with pixel stability when making pixel art games. As before, it relies on using the `rendering/2d/snap/snap_2d_transforms_to_pixel` project setting. If you are making a pixel art game, please test this new release carefully and let us know how it goes.

#### 2D physics interpolation
Fixed timestep a.k.a. physics interpolation is now implemented for 2D ([GH-88424](https://github.com/godotengine/godot/pull/88424)), forward-ported from the version merged for Godot 3.6 last year ([GH-76252](https://github.com/godotengine/godot/pull/76252)).

> This will help address cases of position/camera jitter in 2D games, and should complement some of the pixel-art focused changes made in the 4.3 dev 4 snapshot.

Reviewer: Should there be a word on 3D physics interpolation already being worked on for a future release? This is bound to be a common question after all.

Reviewer 2: ```suggestion
Fixed timestep a.k.a. physics interpolation is now implemented for 2D ([GH-88424](https://github.com/godotengine/godot/pull/88424)), forward-ported from the version merged for Godot 3.6 last year ([GH-76252](https://github.com/godotengine/godot/pull/76252)).

This will help address cases of position/camera jitter in 2D games, and should complement some of the pixel-art focused changes made in the 4.3 dev 4 snapshot.

3D physics interpolation is also in the works for a future Godot release ([GH-92391](https://github.com/godotengine/godot/pull/92391), possibly 4.4.
```

### Editor

#### Fixes for invalid/corrupt scenes
A common complaint among users is that they get locked out of editing a scene because it contained an instance of a different scene that no longer exists (due to being renamed or deleted). Users learned to dread the well-known “Scene invalid/corrupt” error that resulted from this. ([GH-86781](https://github.com/godotengine/godot/pull/86781)) aims to improve the situation by allowing you to open, edit, and fix scenes that have been corrupted due to a missing dependency. This should make the process of refactoring your projects within Godot feel quite a bit safer.

Another bugfix on the GDScript side may also help solve situations where using preload() with cyclic dependencies leads to scenes being flagged as invalid ([GH-85501](https://github.com/godotengine/godot/pull/85501)).

Loading of scenes with corrupted or missing dependencies will no longer be aborted ([GH-85159](https://github.com/godotengine/godot/pull/85159)), allowing you to use and fix such scenes without external tools.

#### Add ufbx for importing .fbx files without FBX2glTF

> This monumental effort by Samuli Raivio and Ernest Lee  incorporates the popular ufbx library into Godot to allow for seamlessly importing .fbx files ([GH-81746](https://github.com/godotengine/godot/pull/81746)). Previously users would have to download the FBX2glTF tool separately and Godot would invoke it to convert .fbx files to glTF files in order to import them. ufbx allows us to avoid this process and import .fbx files directly. Although the ufbx library has gone through extensive rigorous testing, this is a huge change and the .fbx file format is notoriously difficult to work with, so please test this carefully and report any bugs you find.

Reviewer: ```suggestion
This monumental effort by Samuli Raivio and Ernest Lee incorporates the popular ufbx library into Godot to allow for seamlessly importing .fbx files ([GH-81746](https://github.com/godotengine/godot/pull/81746)). Previously users would have to download the FBX2glTF tool separately and Godot would invoke it to convert .fbx files to glTF files in order to import them. ufbx allows us to avoid this process and import .fbx files directly. Although the ufbx library has gone through extensive rigorous testing, this is a huge change and the .fbx file format is notoriously difficult to work with, so please test this carefully and report any bugs you find.
```

- The project manager also got a visual and usability overhaul, with a better layout and a look unified with UI conventions of the editor ([GH-87443](https://github.com/godotengine/godot/pull/87443)). This also introduces initial work on making any kind of network-related feature opt-in in the editor, to give users full control over if and when Godot should communicate with the Internet (e.g. querying the Asset Library for assets, or the Godot website for export templates, etc.).

#### Automatic checking for engine updates
We finally implemented a long-requested feature in the project manager to check for new Godot versions ([GH-75916](https://github.com/godotengine/godot/pull/75916)). This is convenient both when testing pre-release versions, to be notified when beta 1 or future releases are published, but also for new maintenance or feature releases in stable branches.

Out of concern for users’ privacy, this feature is not enabled by default, but can be toggled easily by enabling the “Online” network mode in the project manager’s settings.

> #### PackedByteArrays saved with base64 encoding

Reviewer: ```suggestion
#### PackedByteArrays saved with Base64 encoding
```

> One common annoyance with Godot’s text-based scene/resource format (tscn/tres) is that the serialization of PackedByteArray properties takes a lot of space, leading to inflated file sizes, and noisy diffs. To help with that, we changed the serialization of PackedByteArrays to use base64 encoding, which is more compact, especially for bigger arrays ([GH-89186](https://github.com/godotengine/godot/pull/89186)).

Reviewer: ```suggestion
One common annoyance with Godot’s text-based scene/resource format (tscn/tres) is that the serialization of PackedByteArray properties takes a lot of space, leading to inflated file sizes, and noisy diffs. To help with that, we changed the serialization of PackedByteArrays to use Base64 encoding, which is more compact, especially for bigger arrays ([GH-89186](https://github.com/godotengine/godot/pull/89186)).
```

### Display

#### Wayland support for Linux
It took us just under 10 years to implement, after it was requested back in 2014 and further formalized in 2020: Wayland support is now included in Godot 4.3!

The implementation we merged was a massive undertaking led by Riteo ([GH-86180](https://github.com/godotengine/godot/pull/86180)), spanning 2 years of development with extensive testing and contributions by many others. This built upon previous attempts ([GH-23426](https://github.com/godotengine/godot/pull/23426), [GH-27463](https://github.com/godotengine/godot/pull/27463)) which came at a time where Godot’s architecture wasn’t ready for it yet, notably before the 4.0 split between OS (platform) and DisplayServer responsibilities.

> Included in this effort was the introduction of OpenGL ES driver support for desktop devices ([GH-91466](https://github.com/godotengine/godot/pull/91466)). This is currently limited to linux devices.

Reviewer: This might be worth noting just in case:

```suggestion
Included in this effort was the introduction of OpenGL ES driver support for desktop devices ([GH-91466](https://github.com/godotengine/godot/pull/91466)). This is currently limited to Linux devices (although note that Windows drivers generally do not support OpenGL ES *natively*).
```

For more information on the new Wayland support (uncluding testing instructions), please see the [4.3 dev 3 blog post](https://godotengine.org/article/dev-snapshot-godot-4-3-dev-3/#wayland-support-for-linux).

#### D3D12

> Thanks to the tireless work of randomshaper, Godot now supports the Direct3D 12 rendering API as an optional backend on Windows devices ([GH-70315](https://github.com/godotengine/godot/pull/70315)).

Reviewer: ```suggestion
Thanks to the tireless work of RandomShaper, Godot now supports the Direct3D 12 rendering API as an optional backend on Windows devices ([GH-70315](https://github.com/godotengine/godot/pull/70315)).
```

> Official builds have support for D3D12, but in order to use it you still need to download the DirectX Shader Compiler from Microsoft and copy over the dxil.dll file. 

Reviewer: ```suggestion
Official builds have support for D3D12, but in order to use it, you still need to [download the DirectX Shader Compiler from Microsoft](https://github.com/microsoft/DirectXShaderCompiler/releases/latest) and copy over the `dxil.dll` file to the folder the Godot executable is located in. 
```

> We are still evaluating options for being able to provide a D3D12 support that works out of the box, without a proprietary component. But for now to test things you will have to do this manual step (or compile from source, which does it for you).

Reviewer: ```suggestion
We are still evaluating options for being able to provide a D3D12 support that works out of the box, without a proprietary component. But for now, to test things, you will have to do this manual step (or compile from source, which does it for you).
```

### Platforms

#### Single-threaded web exports
For Godot 4.0, we modernized the engine to make heavier use of multi-threading. The web was the last platform where multi-threading wasn’t a given, but support for the required SharedArrayBuffer feature finally seemed widespread (reaching Safari at last), so we decided to go all in and also make Godot’s Web export multi-threaded by default, solving a number of audio issues we had in Godot 3.

Experience has proven that even though SharedArrayBuffer is supported by all browsers nowadays, the conditions it imposes on the web server that host the games are too difficult to uphold. For people who self-host, it’s easy enough, but for people who distribute their games on publishing platforms like itch.io or CrazyGames, it’s often outside their control. The requirements for SharedArrayBuffer (for security reasons) are also at odds with web game monetization options, such as advertisement or payment processing.

> So we’ve had to change course and do the work to re-add a single-threaded mode to Godot ([GH-85939](https://github.com/godotengine/godot/pull/85939)). The engine can now be compiled with threads=no, which disables all threading use and runs all logic on the main thread.

Reviewer: ```suggestion
So we’ve had to change course and do the work to re-add a single-threaded mode to Godot ([GH-85939](https://github.com/godotengine/godot/pull/85939)). The engine can now be compiled with the `threads=no` SCons option, which disables all threading use and runs all logic on the main thread.
```

The Variant types (like `Vector2`, `Vector3`, `Rect2`, `Transform3D`, etc) now use the `InvariantCulture` by default which means you get uniform formatting regardless of the language of the platform that the game is running on ([GH-89547](https://github.com/godotengine/godot/pull/89547)). This is a breaking change but we thought it was worth it, since these types are not meant to be localized so using the `CurrentCulture` makes little sense.

### GDScript
In terms of new GDScript features, Godot 4.3 may be less stacked than 4.2, but rest assured that the team worked very hard for this release.

In terms of new features, binary tokenization on export has been reintroduced ([GH-87634](https://github.com/godotengine/godot/pull/87634)). This brings back the 3.x functionality to export GDScript files in a binary form, which hides the source code and reduces (a bit) the export size. You can also negatively compare types more naturally with the `is not` operator ([GH-87939](https://github.com/godotengine/godot/pull/87939)).

> We also made built-in type methods and utility functions usable as `Callable` ([GH-82264](https://github.com/godotengine/godot/pull/82264), [GH-86823](https://github.com/godotengine/godot/pull/86823)). This makes it possible to call variadic functions such as `print` with `Callable.callv()` easily, without the need of boilerplate. We also worked on export annotations, such as allowing exported arrays to set property hints for their elements ([GH-82952](https://github.com/godotengine/godot/pull/82952)), we added `@export_storage` to hide exported values from the editor while still exporting the value (useful for plugins!) ([GH-82122](https://github.com/godotengine/godot/pull/82122)), and we created ` @export_custom` for more complex hints or potential future hints ([GH-72912](https://github.com/godotengine/godot/pull/72912)).

Reviewer 3: ```suggestion
We also made built-in type methods and utility functions usable as `Callable` ([GH-82264](https://github.com/godotengine/godot/pull/82264), [GH-86823](https://github.com/godotengine/godot/pull/86823)). This makes it possible to call variadic functions such as `print` with `Callable.callv()` easily, without the need of boilerplate. We also worked on export annotations, such as allowing exported arrays to set property hints for their elements ([GH-82952](https://github.com/godotengine/godot/pull/82952)), we added `@export_storage` to hide the property from the Inspector while still exporting the value ([GH-82122](https://github.com/godotengine/godot/pull/82122)), and we created `@export_custom` for more complex hints or potential future hints ([GH-72912](https://github.com/godotengine/godot/pull/72912)).
```

In terms of bug fixes, the GDScript team fixed lambda hot-reloading ([GH-86569](https://github.com/godotengine/godot/pull/86569)) and fixed out of date errors in depended scripts ([GH-90601](https://github.com/godotengine/godot/pull/90601)). Now GDScript cache will be your friend without restarting the editor!

> A lot of efforts have been made to fix GDScript autompletion ([GH-86667](https://github.com/godotengine/godot/pull/86667), [GH-86554](https://github.com/godotengine/godot/pull/86554), [GH-86111](https://github.com/godotengine/godot/pull/86111), and more). For this, we even introduced a testing system to test for autocompletion regressions! ([GH-86973](https://github.com/godotengine/godot/pull/86973))

Reviewer 3: ```suggestion
A lot of efforts have been made to fix GDScript autocompletion ([GH-86667](https://github.com/godotengine/godot/pull/86667), [GH-86554](https://github.com/godotengine/godot/pull/86554), [GH-86111](https://github.com/godotengine/godot/pull/86111), and more). For this, we introduced a testing system to test for autocompletion regressions ([GH-86973](https://github.com/godotengine/godot/pull/86973)).
```

> A lot of efforts have been made to fix GDScript autompletion ([GH-86667](https://github.com/godotengine/godot/pull/86667), [GH-86554](https://github.com/godotengine/godot/pull/86554), [GH-86111](https://github.com/godotengine/godot/pull/86111), and more). For this, we even introduced a testing system to test for autocompletion regressions! ([GH-86973](https://github.com/godotengine/godot/pull/86973))

Reviewer 4: Likely a typo

```suggestion
A lot of efforts have been made to fix GDScript autocompletion ([GH-86667](https://github.com/godotengine/godot/pull/86667), [GH-86554](https://github.com/godotengine/godot/pull/86554), [GH-86111](https://github.com/godotengine/godot/pull/86111), and more). For this, we even introduced a testing system to test for autocompletion regressions! ([GH-86973](https://github.com/godotengine/godot/pull/86973))
```

### XR
- Godot 4.3 requires a new version of the vendor's plugin to unlock all platform specific functionality.
- Fixed tracking issues in the native mobile interface ([GH-91305](https://github.com/godotengine/godot/pull/91305)).
- Improvements in the core OpenXR loop ([GH-89734](https://github.com/godotengine/godot/pull/89734)), this improves tracking stability and prepares XR for running render processes on a separate thread.
- Added support for composition layers ([GH-89880](https://github.com/godotengine/godot/pull/89880)), this allows 2D viewports to be displayed at much higher quality within an XR environment.
- Improvements to foveated rendering/VRS, passthrough, and various other supporting systems.
- Standardized and enhanced support for Hand/finger tracking, body tracking and face tracking.
- Support for various Meta extensions such as hand models, render models and scene discovery.

### Documentation

> - Add self links to methods, properties, etc. in class reference. For example, this allows you to link directly to a method while reading the description instead of having to scroll up and link the method from the table of methods. ([GH-91537](https://github.com/godotengine/godot/pull/91537)).

Reviewer 3: ```suggestion
- Add self links to methods, properties, etc. in the online class reference. For example, this allows you to link directly to a method while reading the description instead of having to scroll up and link the method from the table of methods. ([GH-91537](https://github.com/godotengine/godot/pull/91537)).
- Add syntax highlighting and a copy button for code blocks in the built-in documentation viewer. ([GH-89263](https://github.com/godotengine/godot/pull/89263), [GH-87363](https://github.com/godotengine/godot/pull/87363)).
- Add support for deprecated/experimental messages. C# also uses the deprecation message when generating bindings. ([GH-81458](https://github.com/godotengine/godot/pull/81458), [GH-88730](https://github.com/godotengine/godot/pull/88730)).
```

- Add option to generate GDExtension docs with `--doctool` ([GH-91518](https://github.com/godotengine/godot/pull/91518)).

### Import

> We have added support for the Quite OK Audio (QOA) format as an optional compression option for WAV files in [GH-91014](https://github.com/godotengine/godot/pull/91014). QOA offers a nice tradeoff between performance, file size, and quality compared to the existing options. It is higher quality than the current WAV compression option (IMA-ADPCM) and slightly smaller while being slightly more CPU intensive to use. Compared to Vorbis and Mp3, it is similar in quality, but slightly larger in size and much less CPU intensive to use.

Reviewer: ```suggestion
We have added support for the Quite OK Audio (QOA) format as an optional compression option for WAV files in [GH-91014](https://github.com/godotengine/godot/pull/91014). QOA offers a nice tradeoff between performance, file size, and quality compared to the existing options. It is higher quality than the current WAV compression option (IMA-ADPCM) and slightly smaller while being slightly more CPU intensive to use. Compared to Vorbis and MP3, it is similar in quality, but slightly larger in size and much less CPU intensive to use.
```

## Changelog

**136 contributors** submitted **466 improvements** for this release. See our [**interactive changelog**](https://godotengine.github.io/godot-interactive-changelog/#4.3-beta1) for the complete list of changes since the 4.3-dev6 snapshot. You can also review [all changes included in 4.3](https://godotengine.github.io/godot-interactive-changelog/#4.3) compared to the previous 4.2 feature release.

This release is built from commit [`a4f2ea91a1bd18f70a43ff4c1377db49b56bc3f0`](https://github.com/godotengine/godot/commit/).

## Downloads

{% include articles/download_card.html version="4.3" release="beta1" article=page %}

**Standard build** includes support for GDScript and GDExtension.

**.NET build** (marked as `mono`) includes support for C#, as well as GDScript and GDExtension.
- .NET build requires .NET SDK 6.0 or later ([.NET 8.0](https://dotnet.microsoft.com/en-us/download/dotnet/8.0) recommended) installed in a standard location.
- To export to Android, .NET 7.0 or later is required. To export to iOS, .NET 8.0 is required.

{% include articles/prerelease_notice.html %}

## Known issues

There are currently no known issues introduced by this release.

> With every release we accept that there are going to be various issues, which have already been reported but haven't been fixed yet. See the GitHub issue tracker for a complete list of [known bugs](https://github.com/godotengine/godot/issues?q=is%3Aissue+is%3Aopen+label%3Abug+).

Reviewer: ```suggestion
With every release, we accept that there are going to be various issues which have already been reported but haven't been fixed yet. See the GitHub issue tracker for a complete list of [known bugs](https://github.com/godotengine/godot/issues?q=is%3Aissue+is%3Aopen+label%3Abug+).
```

- Fixed tracking issues in the native mobile interface ([GH-91305](https://github.com/godotengine/godot/pull/91305)).
- Improvements in the core OpenXR loop ([GH-89734](https://github.com/godotengine/godot/pull/89734)), this improves tracking stability and prepares XR for running render processes on a separate thread.
- Added support for composition layers ([GH-89880](https://github.com/godotengine/godot/pull/89880)), this allows 2D viewports to be displayed at much higher quality within an XR environment.
- Improvements to foveated rendering/VRS, passthrough, and various other supporting systems.
- Standardized and enhanced support for Hand/finger tracking, body tracking and face tracking.
- Support for various Meta extensions such as hand models, render models and scene discovery.

### Documentation

- Add self links to methods, properties, etc. in the online class reference. For example, this allows you to link directly to a method while reading the description instead of having to scroll up and link the method from the table of methods ([GH-91537](https://github.com/godotengine/godot/pull/91537)).

> - Add syntax highlighting and a copy button for code blocks in the built-in documentation viewer. ([GH-89263](https://github.com/godotengine/godot/pull/89263), [GH-87363](https://github.com/godotengine/godot/pull/87363)).

Reviewer 3: ```suggestion
- Add syntax highlighting and a copy button for code blocks in the built-in documentation viewer ([GH-89263](https://github.com/godotengine/godot/pull/89263), [GH-87363](https://github.com/godotengine/godot/pull/87363)).
```
